### DISCUSSION TITLE
Introducing VictorOps for Incident Management

### DISCUSSION THREAD

**Discussion by SamraHanifCareem** [apache/superset#23188] [Ideas]
### Problem Statement:

Since superset is used by many organizations can be smaller or larger as their data analytics tools. There are multiple scenarios in which the user creates their important alerts to get the essential information timely and take necessary actions. However, it doesn't happen as expected because of the alerts getting failed. It could be due to multiple issues such as, query timeout, report schedule refusing to recompute, query got failed, etc. As a result, decreases the productivity of a data analyst to generate outcomes and increases an overhead for the team to identify the cause of its failure and go for the resolution.

### Proposed Solution:

In order to avoid delay in important notifications and introducing auto resolution capability VictorOps Integration can be a great feature to be incorporated. 

### VictorOps (VO):

VictorOps is a real-time incident management and collaboration platform that supports faster triggering and resolution of critical incidents which takes place during the execution of alerts.

The rest endpoint integration can be used, as it allows you to send customised alerts and incident details. They get sent to the Splunk on-Call and after receival, are routed to the respective users/teams for more context. Following fields takes part in incident workflow:

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-14f02082-7fff-b22a-da46-0235b257969f"><div dir="ltr" style="margin-left:0pt;" align="left">

Field | Description
-- | --
message_type | (Required) <br/>Possible Values:- CRITICAL: Triggers an incident- WARNING: May trigger an incident, depending on your settings- ACKNOWLEDGEMENT: Acks an incident- INFO: Creates a timeline event but does not trigger an incident- RECOVERY: Resolves an incident
entity_id | Should be unique and remain same throughout the life cycle
entity_display_name | Display name in the UI
state_message | Verbose message field that incorporates long explanation of the problem

</div></b>

For more information on different fields that Splunk On-Call uses please follow the [VictorOps Glossary](https://help.victorops.com/knowledge-base/incident-fields-glossary/#glossary-of-fields.).

They can be sent from any source via an HTTP Post request and the format used for the payload is JSON. _Fig 01 depicts the execution flow of alerts._

![Alert Notification Flow drawio (2)](https://user-images.githubusercontent.com/106157603/221158593-8c2c250d-165f-468e-bd0f-9f87fd699d6a.png)

> _Fig 01: Execution of Alerts_

Some of the possible errors that should be routed to VictorOps for faster resolution:
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-95e9a5bd-7fff-9863-ec69-99b226dbedb1"><div dir="ltr" style="margin-left:0pt;" align="left">

Alert found an error while executing a query
--

- A timeout occurred while executing a query
- A timeout occurred while generating a csv
- Alert on grace period
- Report schedule is still working, refused to recompute
- Report schedule unexpected error

</div></b>

### VO Configuration:

1.  Login to [VictorOps](https://portal.victorops.com/)
2.  Go to Integrations >> 3rd Party Integrations >> REST – Generic.
3.  Get the URL to notify
4.  Generate the routing key by clicking on Setting >> Routing Keys
5. After following the above mentioned steps, the rest endpoint will look like this:  https://alert.victorops.com/integrations/generic/20131114/alert/xxxxxxxxxxxxxx/$routing_key

### VO Integration with Superset:

Following changes are expected in Alerts and Reports for the integration of VO with superset:

1.  Adding VO_Endpoint as a configuration parameter in config.py.
2. Adding VO as a new notification method. Please refer to _Fig 02 depicting Alerts and Reports User Interface._
3.  Taking routing_key as an input from the user.
4.  Storing the accepted routing_key in the database.
5.  catching the errors on which the incident should be raised For e.g. Alert Query Timeout, Sql Query Failed, Report schedule refused to re-compute, etc.
6.   Raise incident.
7.  Getting necessary payload fields as discussed in [VO documentation](https://help.victorops.com/knowledge-base/rest-endpoint-integration-guide/#recommended-rest-endpoint-integration-fields).
8.  Integrating the generated endpoint along with the routing_key provided by the user.

<img width="1412" alt="VO UI" src="https://user-images.githubusercontent.com/106157603/221158631-c1864c3c-e070-429f-b01d-4c13720c55a7.png">

> _Fig 02: Alerts and Reports User Interface_

### New or Changed Public Interfaces:

-   There is a minor proposed changes in the public interface as depicted in Fig 02.
-   New column needs to be added in the alert/report model.

### New dependencies:
 None
### Migration Plan and Compatibility:

- The required PR needs to be merged after the development

### Rejected Alternatives:
N/A

**Comment by mdeshmu** [comment]
First of all thanks a ton @SamraHanifCareem for putting great effort to write and initiate this discussion. Here are my thought on what is proposed so far:

I really like the use case listed in section "Some of the possible errors that should be routed" but overall proposal is too restrictive. 
The scope of the discussion is limited to integrating VictorOps REST endpoint and proposal is to hard-code to VictorOps specific glossary. This is going to be helpful to limited organizations only as not all organizations use VictorOps. 

Scope should be wide enough to configure triggers and having those fire off a web-hook, rather than configuring any specific/particular service integrations. The inputs should be generic enough & applicable to any web-hook. Consider an example below:

![image](https://user-images.githubusercontent.com/57723564/221184639-34637a02-1eb3-4630-8153-1a442b526a72.png)

Our Organization not only wants to create an incident but in some cases we want to take corrective action automatically. 
For example: 
* Calling a Superset API to re-trigger an action
* Calling a internal API to trigger an internal workflow

Hence keeping the integration generic would really help us and many others. We can send to a chat web-hook like Microsoft Teams or send to a web-hook of any other incident management tool. Integration should be able to fire web-hook for both success or failure of an alert or report. 

If sounds feasible, Existing Jinja context Macros can be made available to the payload (If not available already) and there is room to add support for new Macros specific to alert & report modal. 

Question: How are we going to determine whether an alert or report has failed to send the notification ?

**Reply by SamraHanifCareem** [reply]
@mdeshmu Currently, we are catching the errors from celery workers and displaying them to the user in _Execution Logs._ We can do the same for other notification methods. 

For instance, let us take an example of unauthorised access to VictorOps or any other web-hook, the post api will return 401 and some response containing the detail of the error message. We can catch the error and update it to the Execution Log. 

In Addition, the failure messages can also be send to email (if exist) similar to which we are doing for other notification methods. For instance, the slack api token error as mentioned below:

<img width="1438" alt="Screenshot 2023-03-22 at 11 26 33 PM" src="https://user-images.githubusercontent.com/106157603/227002388-2a9c5181-b303-4a34-989b-363535d0513b.png">

**Comment by eschutho** [comment]
Thank you for this discussion, and I agree with @mdeshmu that we would want to keep any solutions flexible for all users. Currently the Celery workers are raising errors and passing them to your logger as well as to a statsd system. You can configure either of these tools to your liking. What most installations will do is then set threshold levels to monitor the logs for numbers of errors or statsd counts for error or success thresholds and then integrate with PagerDuty or some other system for alerting the team when those thresholds are reached. For retries, Celery should be able to handle any failed jobs that did not complete. 

The alert feature is closely tied to database queries, rather than Superset system performance. If you would like to have a db query that calls an api other than slack when a condition has been met, then that sounds like a different use case than what you're looking for?

**Reply by mdeshmu** [reply]
@SamraHanifCareem do you plan  to implement and contribute the generic approach? 
If not, I would like to initiate an internal project with my firm to implement the generic approach and then later contribute to the community.

**Reply by SamraHanifCareem** [reply]
@mdeshmu Let me do some in-depth research on this and get back to you in a couple of days

**Comment by mdeshmu** [comment]
@eschutho @rusackas I would like to initiate an internal project within my firm to implement the generic approach that we discussed above and we would later contribute to the community. Basically we want to add new notification web-hook channel  in addition to existing email and slack.

What is the appropriate way to communicate and finalize the specifications of the implementation? Who would be appreciate point of contact from Superset PMC?

**Reply by rusackas** [reply]
I would say you can bring this up at the next Superset Town hall if you'd like... If I'm parsing this thread correctly, I'm doubtful that the webhook implementation will be seen as controversial, as long as the code is kept DRY and there are sufficient tests. I'd say go ahead and implement this, and then open a PR to upstream the feature, so we can hash out the details there.
